### PR TITLE
Retry timeout

### LINKED ISSUE #660: PVF: Add a new class of "possibly invalid" errors
# ISSUE

## Overview

PVF execution plays a critical role in our dispute process. If we execute a candidate block and find it is invalid, we vote against it. If some validators' votes do not agree, a dispute is initiated.

Now, in the real world we may try to execute a candidate and it fails due to a hardware fault, operator error, some rare bug, etc. In this case, voting against would initiate a dispute, which is not what we want -- it may get the validator slashed!

So, right now we have two main failure states:

1. `InvalidCandidate`, where we always vote against.
1. `InternalError`, where we never vote against, but do retry once (since https://github.com/paritytech/polkadot/pull/7011) to let transient error conditions clear.

And we do actually have a third state: `InvalidCandidate::AmbiguousWorkerDeath`, which happens when the worker process dies for an unknown reason. In this case we retry once, but if it still fails, then we **do** vote against. This would happen if the PVF execution takes up a large amount of memory, causing OOM, in a way that is reproducible.

## Proposal

So, I am proposing that we extend this third category, giving it a catch-all name like `PossiblyInvalid`. We would retry these, and only on continued failure deem the candidate invalid.

Other errors that we could treat this way:

1. **Timeout due to exceeded CPU time.** Currently this always results in `InvalidCandidate`. But, under conditions of high local load we may do very little work, while still counting CPU time, eventually timing out for a valid candidate. This can be retried after a delay, in the hopes that the load died down. We have to be careful though. It may lead to yet more increase to load, on the other hand we may prevent disputes by retrying! 
  a. (We may also want to detect conditions of high load and deal with it somehow.)
  b. NOTE: There may not be enough time for a retry in candidate validation from backing.
1. We can treat **`RuntimeConstruction`** this way. Right now it always results in Invalid, though it should be a local issue. https://github.com/paritytech/polkadot-sdk/issues/661 would be the full fix, but in the meantime it would make sense to retry in this case before voting against.
1. Any others?

Having this separate variant would also just make the code easier to reason about, and easier to add more "retry-then-vote-against" cases in the future.

Also, it would be nice to make these three distinct categories clear in the documentation!

### PR DESCRIPTION
**Pull request by eagr**
closes #660 

Is the retry delay being used (200ms) good enough for the retry for job timeout?

Btw, how serious are you with the `SchrodingersCandidate`? I got zero problem with the name. :)) @mrcnski @s0me0ne-unkn0wn

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4429721

**Comment by mrcnski** [comment]
> Btw, how serious are you with the `SchrodingersCandidate`? I got zero problem with the name. :)) @mrcnski @s0me0ne-unkn0wn

I love that you asked this. 😂 I vote "yes", only question is, should we include the umlaut?

File: polkadot/node/core/pvf/src/execute/queue.rs [hunk 1]
```diff
@@ -361,6 +357,10 @@ fn handle_job_finish(
 			Err(ValidationError::PossiblyInvalid(PossiblyInvalidError::AmbiguousJobDeath(err))),
 			None,
 		),
+		// Either the worker or the job timed out. Kill the worker in either case. Treated as
+		// definitely-invalid, because if we timed out, there's no time left for a retry.
```
Comment (mrcnski) [inline polkadot/node/core/pvf/src/execute/queue.rs:361]: Could be reworded as it's potentially confusing. It is only definitely-invalid in backing where we don't retry the possibly-invalid candidates. In backing there's no time for a retry (and it's stricter).

**Comment by mrcnski** [comment]
Hmm, actually I'm not sure if we should retry, at least with the current timeout which is `DEFAULT_APPROVAL_EXECUTION_TIMEOUT = 12`. That's already a long time given for execution.

One idea is to allow a retry but cut that timeout in half. The idea is that if execution is taking 6 seconds (which is already a long time) then something went wrong (i.e. maybe the worker/job communication broke down).

And speaking of communication, the host/worker comms (note, not worker/job comm) have a really long timeout (like factor of 4 of the default timeout). We shouldn't retry if that fails because there'd just be no time left. But that long timeout needs to remain long, because it's wall clock time, whereas in the job we timeout based on CPU clock time. (That probably sounds really confusing, so see https://github.com/paritytech/polkadot/issues/4217.)

But getting back to cutting the approval timeout in half, I wonder if opens up some attacks, like candidates getting through backing but timing out on some validators with slower machines, leading to disputes. 6 seconds would already be 3x the backing timeout, but I can see it happening. cc @eskimor

**Comment by eagr** [comment]
You have a change of mind now or did I misunderstood your [comment](https://github.com/paritytech/polkadot-sdk/pull/2406#issuecomment-1818810225)?

12s is a really long time. How often does a timeout happen on a valid candidate in practice? If it's extremely rare, I suppose we shouldn't. In that case, I'm happy to close this as it no longer makes sense to switch the category.

**Comment by mrcnski** [comment]
You understood correctly, I just forgot that we had bumped the approval timeout to 12s.

> How often does a timeout happen on a valid candidate in practice?

Not sure, I suppose my concerns are more theoretical in nature. Like if there is a validator that is 3x slower than the rest, it can potentially be exploited to create disputes. I think in practice it's more likely that a timeout happened due to some local conditions, like some deadlock, comm bug, etc... though still not very likely.

Maybe some compromise is possible, like 8s + a retry? Not sure, @eskimor would have a better idea.

**Comment by eskimor** [comment]
Definitely no to shortening the timeout. We have measured fluctuations on the same machine up to a factor of 3. An additional factor of 2 does not seem to be too conservative, given that validators don't run identical machines.

Whether a retry on timeout makes sense is an interesting question:

1. We would be too slow to vote and thus approval voting would escalate because of our no-show. This is an issue, as without the retry we would have raised a dispute - which is a full escalation, but in that case at least someone will pay the bill.
2. Retrying increases the chances of us voting valid on a candidate. If a super majority finds this borderline candidate to be invalid, we just increased our chances of getting slashed significantly. Although if everybody does this, then the chances of a super majority finding the candidate invalid also gets smaller. On top of this with https://github.com/paritytech/polkadot-sdk/issues/635 in place, collectively it does not matter much if 1 or 10 validators voted valid on an invalid candidate.

2 is not much of a reason, but 1 means it is free to trigger an escalation (and break sharding). So retry on timeout is not an option ... in fact if we had the time, I don't see why we would ever retry instead of just increasing the timeout to begin with. What would be the value of retrying, instead of increasing the timeout? If you restart - any work already done (even if little) would go to waste. While with a longer timeout you keep it and have a better success chance, given the same time constraints.

**Comment by eskimor** [comment]
> 12s is a really long time. How often does a timeout happen on a valid candidate in practice? 

It is.

>  If it's extremely rare

We at least hope so. That's the whole point of that long timeout. We actually should increase the noshow timeout because of above reasoning, as execution time alone is not the full story. 

For handling dishonest backers, we need [time disputes](https://github.com/paritytech/polkadot-sdk/issues/742#issuecomment-1691884835).

**Comment by eskimor** [comment]
More to the point, why we really don't want retry on timeout: If backers ignore their 2s timeout, they are already exploiting network resources free of charge. Even, without escalation, by retrying we would then do this twice - doubling the cost.

**Comment by mrcnski** [comment]
> We have measured fluctuations on the same machine up to a factor of 3.

Wow, a 3x fluctuation on the *same machine*? This must be wall clock time? The timeouts in question are CPU time. But still, I agree that we should have a large margin for differently-performing validators.

> in fact if we had the time, I don't see why we would ever retry instead of just increasing the timeout to begin with. What would be the value of retrying, instead of increasing the timeout?

If execution is taking that long, then it could be due to some deadlock etc. Waiting a second and then retrying could avoid some local nondeterminism. Then again, I can't think of any specific internal scenarios that would cause this to occur, except for high load on the machine, in which case retrying could make things worse.

Maybe, there is some way for the node to query whether the execution job is at least *making progress* as opposed to spinning in a loop. But I might be getting into the weeds here.

**Comment by eagr** [comment]
Thanks for the explanation and pointers! @eskimor Those are so informative. Curious, what difference does it make if we no-show for longer time, if any? Say we double the timeout.

You could just close this if there's no reason to proceed. :) @mrcnski

**Comment by mrcnski** [comment]
> If polkavm becomes twice as fast

Looks like it [just happened](https://forum.polkadot.network/t/announcing-polkavm-a-new-risc-v-based-vm-for-smart-contracts-and-possibly-more/3811/33).

**Comment by mrcnski** [comment]
Closing. Thanks for the hustle @eagr and apologies for the spurious suggestion! Will also close the related issue as there's nothing left to do.

**Comment by mrcnski** [comment]
> Maybe, there is some way for the node to query whether the execution job is at least making progress as opposed to spinning in a loop.

I suppose it can be done in PVM by checking if gas is going up. Maybe overly paranoid. I wouldn't do it unless timeouts become a source of issues.

**Pull request closed without merging**